sycl/test/e2e_test_requirements/check-correctness-of-requires.cpp

```diff
@@ -0,0 +1,10 @@
+// This test checks that all "REQUIRES" strings contain the right feature names
+// If this test fails:
+// 1. there is some typo/non-existing feature request in the
+//    modified test.
+// 2. ...or, there is some new feature. In this case please update the set of
+//    features in check-correctness-of-requires.py
+//
+// RUN: grep -rI --include=*.cpp "REQUIRES: " %S/../../test-e2e | sed -E 's|.*/test-e2e/||' > %t
```

Comment:
Please, don't use grep! It's error prone. We must leverage `llvm-lit` scripts to get the list of features.

Reply:
I wonder if we can write a custom format, which would automatically read requirements for every test and compare them with reference. Something like I've done for checking that every header can be included standalone [here](https://github.com/intel/llvm/blob/sycl/sycl/test/format.py).

I don't even think that we would need to redefine discovery step, we just need to redefine the test step: instead of running any commands from the file we would just check requirements and fail if we encountered anything unknown. The only thing which we will need is to somehow make `lit` discover tests in a different directory for us.

Reply:
> > Please, don't use grep! It's error prone
> 
> Just out of curiosity, what sort of errors can we run into with grep?

This script already has a bug. llvm-lit is configured to consider `.c` files as test files as well, but there might be other file extension for the tests in test-e2e directory. 
Someone can use "REQUIRES: " in the manner that llvm-lit ignores (e.g. in the middle of the comments).

We rely on some logic of llvm-lit scripts, but we don't leverage the source code of these scripts as all.

Reply:
> I wonder if we can write a custom format, which would automatically read requirements for every test

This part is already done in the llvm-lit.

  --show-used-features  Show all features used in the test suite (in XFAIL, UNSUPPORTED and REQUIRES) and exit

> llvm-lit --show-used-features llvm/test/

> `aarch64-registered-target amdgpu-registered-target arm-registered-target asserts avr-registered-target bpf-registered-target can-execute csky-registered-target curl cxx-shared-library debug debug_frame default_triple diasdk disabled do-not-run-me ed. examples exegesis-can-execute-aarch64 exegesis-can-execute-mips exegesis-can-execute-x86_64 exegesis-can-measure-latency exegesis-can-measure-latency-lbr expensive_checks fma3 has_logf128 have_tf_aot have_tflite hexagon-registered-target host-byteorder-little-endian host={{.*-windows-msvc}} httplib intel-jitevents intel_feature_xe lanai-registered-target ld_emu_elf32ppc libcxx-used libxml2 lld llvm-64-bits llvm-driver llvm-dylib llvm_inliner_model_autogenerated llvm_raevict_model_autogenerated loongarch-registered-target mips-registered-target msan msp430-registered-target native non-root-user nvptx-registered-target object-emission plugins powerpc-registered-target reverse_iteration riscv-registered-target riscv64-registered-target shell sparc-registered-target spirv-tools static-libs system-aix system-darwin system-freebsd system-linux system-netbsd system-windows system-zos systemz-registered-target target-byteorder-little-endian target-x86_64 target=aarch64-pc-windows-{{.*}} target=aarch64{{.*}} target=amdgcn-{{.*}} target=arm{{.*}} target=avr{{.*}} target=hexagon-{{.*}} target=loongarch{{.*}} target=nvptx{{.*}} target=powerpc64-unknown-linux-gnu target=powerpc64{{.*}} target=powerpc{{.*}} target=r600{{.*}} target=riscv{{.*}} target=sparc{{.*}} target=target=powerpc64-unknown-linux-gnu target=x86_64-{{.*-ps[45]}} target=xcore{{.*}} target={{(aarch64|arm).*}} target={{(i686|i386).*}} target={{(mips|mipsel)-.*}} target={{.*-(cygwin|windows-gnu|windows-msvc)}} target={{.*-(cygwin|windows-msvc|windows-gnu)}} target={{.*-linux.*}} target={{.*-windows-(gnu|msvc)}} target={{.*-windows.*}} target={{.*windows.*}} target={{.*}} target={{.*}}-aix{{.*}} target={{.*}}-apple{{.*}} target={{.*}}-windows-gnu target={{.*}}-zos{{.*}} target={{i686.*windows.*}} thread_support to-be-fixed use_msan_with_origins uses_COFF vc-rev-enabled ve-registered-target vg_leak webassembly-registered-target x86-registered-target x86_64-apple x86_64-linux xar zlib zstd`

Reply:
> This script already has a bug. llvm-lit is configured to consider .c files as test files as well, but there might be other file extension for the tests in test-e2e directory.

We have previously agreed with Nikita that treating `.c` files as E2E tests was a bug and it should be fixed. We don't have any `.c` tests and we can compile SYCL code (which requires C++17) in C mode. But in general I tend to agree, there is a request to introduce `.sycl` file extension, for example.

Reply:
@bader thanks! `llvm-lit sycl/test-e2e --show-used-features` works good, I'll try it.

Reply:
A note here, `--show-used-features` does not show features added to requires/unsupported/xfail from lit.local.cfg files.